Use IFileService to save macro partial views in package migration

ronaldbarendse opened the pull request:
### Prerequisites

- [x] I have added steps to test this contribution in the description below

### Description
When a package migration tried to save macro partial views from an embedded package.xml file (like in The Starter Kit), it directly used `File.WriteAllText()` instead of the `IFileService.SavePartialViewMacro()`. If the `Views/MacroPartials` directory didn't exist yet, this also caused a `DirectoryNotFoundException` to be thrown (which would cause a `BootFailedException` when the migrations are automatically run).

Testing steps:
1. Before applying this PR on a clean install, remove the `Views/MacroPartials` directory and try to install The Starter Kit (`dotnet add package Umbraco.TheStarterKit --version 9.0.0-rc002`): this will throw an exception while booting up;
2. Manually create the `Views/MacroPartials` directory and start the site: the package migration will now successfully run and the `FeaturedProducts.cshtml` and `LatestBlogposts.cshtml` files should be created in this directory;
3. Apply this PR to a clean install and repeat the first step: it should now run the package migration without issues and the macro partial view files should also exist.

bergmania commented:
When testing your PR. I get this error on the starterkit frontpage:
![image](https://user-images.githubusercontent.com/1561480/133754788-6bb874e1-8bd4-4de8-b9b9-458bf019d9b0.png)

```
System.InvalidOperationException: The view 'Components//~/Views/MacroPartials/LatestBlogposts.cshtml' was not found. The following locations were searched:
~/Views/MacroPartials/LatestBlogposts.cshtml
   at Microsoft.AspNetCore.Mvc.ViewEngines.ViewEngineResult.EnsureSuccessful(IEnumerable`1 originalLocations)
   at Microsoft.AspNetCore.Mvc.ViewComponents.ViewViewComponentResult.ExecuteAsync(ViewComponentContext context)
   at Microsoft.AspNetCore.Mvc.ViewComponents.ViewViewComponentResult.Execute(ViewComponentContext context)
   at Umbraco.Cms.Web.Common.Macros.PartialViewMacroEngine.Execute(MacroModel macro, IPublishedContent content)
   at Umbraco.Cms.Web.Common.Macros.MacroRenderer.<>c__DisplayClass24_0.<ExecuteMacroOfType>b__0()
   at Umbraco.Cms.Web.Common.Macros.MacroRenderer.ExecuteProfileMacroWithErrorWrapper(MacroModel macro, String msgIn, Func`1 getMacroContent, Func`1 msgErr)
   at Umbraco.Cms.Web.Common.Macros.MacroRenderer.ExecuteMacroWithErrorWrapper(MacroModel macro, String msgIn, String msgOut, Func`1 getMacroContent, Func`1 msgErr)
   at Umbraco.Cms.Web.Common.Macros.MacroRenderer.ExecuteMacroOfType(MacroModel model, IPublishedContent content)
   at Umbraco.Cms.Web.Common.Macros.MacroRenderer.RenderAsync(MacroModel macro, IPublishedContent content)
   at Umbraco.Cms.Web.Common.Macros.MacroRenderer.RenderAsync(String macroAlias, IPublishedContent content, IDictionary`2 macroParams)
   at Umbraco.Cms.Core.Templates.UmbracoComponentRenderer.RenderMacroAsync(IPublishedContent content, String alias, IDictionary`2 parameters)
   at Umbraco.Cms.Core.Templates.UmbracoComponentRenderer.RenderMacroAsync(Int32 contentId, String alias, IDictionary`2 parameters)
   at Umbraco.Cms.Web.Common.UmbracoHelper.RenderMacroAsync(String alias, IDictionary`2 parameters)
   at AspNetCore.Views_Partials_grid_editors_macro.ExecuteAsync() in C:\Users\Bjarke\TestRonald\Views\Partials\grid\editors\macro.cshtml:line 13
   at Microsoft.AspNetCore.Mvc.Razor.RazorView.RenderPageCoreAsync(IRazorPage page, ViewContext context)
   at Microsoft.AspNetCore.Mvc.Razor.RazorView.RenderPageAsync(IRazorPage page, ViewContext context, Boolean invokeViewStarts)
   at Microsoft.AspNetCore.Mvc.Razor.RazorView.RenderAsync(ViewContext context)
   at Microsoft.AspNetCore.Mvc.ViewFeatures.HtmlHelper.RenderPartialCoreAsync(String partialViewName, Object model, ViewDataDictionary viewData, TextWriter writer)
   at Microsoft.AspNetCore.Mvc.ViewFeatures.HtmlHelper.PartialAsync(String partialViewName, Object model, ViewDataDictionary viewData)
   at AspNetCore.Views_Partials_grid_editors_base.ExecuteAsync() in C:\Users\Bjarke\TestRonald\Views\Partials\grid\editors\base.cshtml:line 6
```

bergmania commented:
Also the MacroPartialViews are saved incorrectly

![image](https://user-images.githubusercontent.com/1561480/133755064-1d2c9bb0-d5d9-4348-a5cf-b228f8bf3145.png)

bergmania commented:
Not sure how to best solve this.. 
- IMO the packages should not be created with the full path, but only the relative path after `~/Views/MacroPartials`. But if we assume this, we will break existing packages. So we should most likely handle both cases in the import feature

ronaldbarendse commented:
@bergmania It's indeed inconsistent, as this todo comment points out: https://github.com/umbraco/Umbraco-CMS/blob/f68717c1c8c7dfe994f5c8931052c9c8b793e2b3/src/Umbraco.Web.BackOffice/Controllers/MacrosController.cs#L318-L322

While selecting/picking a macro partial view, it can come from the `Views\MacroPartials` directory (which is editable from the tree and is using the `MacroPartialsFileSystem`) or from `App_Plugins\*\Views` (provided by an external package). We should only include the ones from `Views\MacroPartials` in the exported XML, as the others would/should already be included in a package (as NuGet content file).

I've fixed this by only exporting macro partial views from the `Views\MacroPartials` directory and removing the prefix. For backwards compatibility, the prefix is also removed while importing the package XML and macro partial views outside of the MacroPartialsFileSystem are now skipped (a warning is logged when this happens).

ronaldbarendse commented on src/Umbraco.Core/Packaging/PackagesRepository.cs:
```diff
@@ -414,9 +414,10 @@ private void PackageMacros(PackageDefinition definition, XContainer root)
-            IEnumerable<string> views = packagedMacros.Select(x => x.MacroSource).Where(x => x.EndsWith(".cshtml"));
-            PackageMacroPartialViews(views, root);
+            // Get the partial views for macros and package those (exclude views outside of the default directory, e.g. App_Plugins\*\Views)
+            IEnumerable<string> views = packagedMacros.Where(x => x.MacroSource.StartsWith(Constants.SystemDirectories.MacroPartials))
+                .Select(x => x.MacroSource.Substring(Constants.SystemDirectories.MacroPartials.Length).Replace('/', '\\'));
```
This does the filtering on and removing of the prefix. It also replaces forward slashes to backward slashes, as the virtual path is constructed differently compared to the normal views, so this aligns everything a bit more (both are handled correctly though).

ronaldbarendse commented on src/Umbraco.Infrastructure/Packaging/PackageDataInstallation.cs:
```diff
@@ -1280,29 +1255,46 @@ public IReadOnlyList<ILanguage> ImportLanguages(IEnumerable<XElement> languageEl
-                // TODO: Do we overwrite? IMO I don't think so since these will be views a user will change.
-                if (!System.IO.File.Exists(physicalPath))
+                // Remove prefix to maintain backwards compatibility
+                if (path.StartsWith(Constants.SystemDirectories.MacroPartials))
                 {
-                    System.IO.File.WriteAllText(physicalPath, contents, Encoding.UTF8);
+                    path = path.Substring(Constants.SystemDirectories.MacroPartials.Length);
+                }
+                else if (path.StartsWith("~"))
+                {
+                    _logger.LogWarning("Importing macro partial views outside of the Views/MacroPartials directory is not supported: {Path}", path);
+                    continue;
+                }
```
This is only required to maintain backwards compatibility with package XMLs that still contain the virtual path.

ronaldbarendse commented on src/Umbraco.Infrastructure/Packaging/PackageInstallation.cs:
```diff
@@ -44,20 +44,29 @@ public InstallationSummary InstallPackageData(CompiledPackage compiledPackage, i
+
+            packageDefinition.ContentNodeId = installationSummary.ContentInstalled.FirstOrDefault()?.Id.ToInvariantString();
+            
+            foreach (var x in installationSummary.MediaInstalled)
+                packageDefinition.MediaUdis.Add(x.GetUdi());
```
I'm not sure where the updated package definition is used, but I've added media and removed the stylesheets as they aren't tracked by ID.

So the following types aren't currently tracked, because that would require serializing them to XML (that includes the path and file contents):
- `MacroPartialViewsInstalled`
- `StylesheetsInstalled`
- `ScriptsInstalled`
- `PartialViewsInstalled`

ronaldbarendse replied:
The package definition is now also updated with the imported stylesheets, scripts and partial views paths. Macro partial views are part of the macro, so isn't defined as separate entity in the package definition, which is fine I guess.
